lib/dependency_validator.dart

```diff
@@ -57,6 +63,29 @@ void run() {
 
   logger.info('Validating dependencies for $packageName\n');
 
+  // `loadPackageConfig` (used below) relies on having a file built from an absolute URI.
+  final packageConfig = File.fromUri(Uri.file('${Directory.current.path}/.dart_tool/package_config.json'));
+
+  // URIs in packageConfig may be relative to its directory. Change the working directory to account for this.
+  final rootDir = Directory.current;
+  final dir = Directory('${rootDir.path}/.dart_tool');
+  Directory.current = dir;
+
+  // Find packages that provide executables.
+  final packagesWithExecutables = Set<String>();
+  for (final package in (await loadPackageConfig(packageConfig)).packages) {
+    final path = package.root.path;
+    if (Directory('$path${path.endsWith('/') ? '' : '/'}bin').existsSync()) {
+      packagesWithExecutables.add(package.name);
+    }
+  }
+
+  // Set working directory back.
+  Directory.current = rootDir;
+
+  logDependencyInfo(
+      'the following packages contain executables, they are assumed to be used:', packagesWithExecutables);
```

Comment:
I think this is going to be too noisy since above we're tracking all transitive dependencies and not just those that are directly depended on by this package. In addition to only checking direct dependencies for executables (instead of all transitive ones, as well), I think we should filter this list to be the packages that are being considered "used" _only_ because they contain executables and not for any other reason. That way this info log can be an indicator to maintainers that the listed packages might be able to be removed if their executables aren't being used.